### ISSUE TITLE
metadata: Metadata association problems with children processes

### ISSUE THREAD

**Issue by kakkoyun** [parca-dev/parca-agent#1157]
Metadata association is straightforward right now (just matching exact PIDs). It causes problems associating the relevant information from service discovery (e.g. kubernetes, systemd). We need to attach all the metadata information to the processes within the encapsulator (cgroup, container, or pod). See the example below.

![](https://cdn.discordapp.com/attachments/877547706334199821/1055056854054621214/Screenshot_2022-12-21_at_10.34.13.png)

```
$ ps auxww
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
rabbitmq       1  0.0  0.0   2616  1668 ?        Ss   09:01   0:00 /bin/sh /opt/rabbitmq/sbin/rabbitmq-server
rabbitmq      11 17.2  2.6 4496400 439992 ?      Sl   09:01   9:55 /usr/local/lib/erlang/erts-12.3.2/bin/beam.smp -W w -MBas ageffcbf -MHas ageffcbf -MBlmbcs 512 -MHlmbcs 512 -MMmcs 30 -P 1048576 -t 5000000 -stbt db -zdbbl 128000 -sbwt none -sbwtdcpu none -sbwtdio none -JPperf true -B i -- -root /usr/local/lib/erlang -progname erl -- -home /var/lib/rabbitmq -- -pa  -noshell -noinput -s rabbit boot -boot start_sasl -syslog logger [] -syslog syslog_error_logger false -kernel prevent_overlapping_partitions false
```

We should have associated the metadata with the given process for this particular case.
https://github.com/parca-dev/parca-agent/blob/5a538215709969388c2fa31441bcad06e9d7b2e9/pkg/discovery/kubernetes.go#L127

However, we rely on Kubernetes API to deliver updates and heavily on caching in this path. It's possible that we might have missed some updates. It needs further investigation.

For replication:
```shell
kubectl apply -f https://github.com/rabbitmq/cluster-operator/releases/latest/download/cluster-operator.yml
kubectl apply -f https://raw.githubusercontent.com/rabbitmq/cluster-operator/main/docs/examples/hello-world/rabbitmq.yaml
```
* https://github.com/docker-library/rabbitmq/blob/master/3.11/ubuntu/Dockerfile
* https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/scripts/rabbitmq-server

**Comment by kakkoyun** [comment]
Could be related https://github.com/parca-dev/parca-agent/issues/688

**Comment by kakkoyun** [comment]
I'm currently working on this.

**Comment by kakkoyun** [comment]
Fixed by #1232
